### ISSUE TITLE
Run error！

### ISSUE THREAD

**Issue by 437861619** [yunjuanwang/robust_learning_for_data_poisoning_attack#1]
from art.classifiers import KerasClassifier；
from art.classifiers import PyTorchClassifier；
These two lines of code report an error，The file 'art.classifiers' cannot be found.

**Comment by yunjuanwang** [comment]
The issue is because the adversarial-robustness-toolbox package has been modified. I change these two lines as follows:
from art.estimators.classification import KerasClassifier, PyTorchClassifier

Please let me know if you have more questions.

**Comment by 437861619** [comment]
Thank you for your timely clarification!
Then, the art package seems to require a tensorflow environment to run.
art\attacks\poisoning\backdoor_attack_dgm_red.py", line 39 Prompt for an error :'No module named 'tensorflow''.

**Comment by yunjuanwang** [comment]
One way you can do is: pip install tensorflow or pip install adversarial-robustness-toolbox tensorflow.
Please see https://github.com/Trusted-AI/adversarial-robustness-toolbox/wiki/Get-Started#setup for more information regarding how to use art package.

**Comment by 437861619** [comment]
I would like to ask you about the version of pytorch you are using, the one I am using has some conflicts

**Comment by yunjuanwang** [comment]
I'm using torch 1.7.0

**Comment by 437861619** [comment]
Thank you for your answer, I have successfully generated poisoned data, but the results of the experiment are not very clear, is there a lack of code here to generate the graph of the experiment results, like in the paper.

**Comment by yunjuanwang** [comment]
If you'd like to regenerate figure 3 in the paper, you need to run mnist_ntk.py and cifar_ntk.py multiple runs with different network widths and different regimes. You then record the test accuracy for different parameters which is needed for generating the figure.

**Comment by 437861619** [comment]
Okay, do you know how to run a python program in a loop with different parameters so that you don't have to change them multiple times, and how to get the accuracy on the clean data set?

**Comment by 437861619** [comment]
Ok, I'm already organizing the data, so how do I get the results of your other experimental results graphs like Figure 2.

**Comment by yunjuanwang** [comment]
I upload a plot_figure.ipynb file for your reference. You need to put the test accuracy that you get for different parameters into the corresponding variables. I strongly suggest that for each parameter, you run the experiment at least 20 times. I also suggest you record the result into a file, which is convenient for you to import and plot the figure.

**Comment by 437861619** [comment]
Thank you very much for your patient answer!

**Comment by 437861619** [comment]
I'm sorry to bother you, and I have a few questions about the paper:
1.Is the meaning of per-sample corruption budget B as the maximum value of the corruption budget?
2.What are the marginal parameters γ and linear separator and the definition of the O function?
3.With respect to Figure 2, what is the meaning of the vertical coordinates, and what conclusion are you trying to prove?
4.In Figure 3, why is the accuracy of regime C higher on the MNIST dataset and lowest on the CIFAR10 dataset?
5.From which experimental results do you fully illustrate the conclusion that SGD is robust to poisoning attacks?
Also, I would like to get your contact information to reach you outside of GitHub, thank you very much!

**Comment by yunjuanwang** [comment]
Reply:
1. Per-sample corruption budget is the maximum L2 norm perturbation you can add to each training sample.
2. gamma is the margin in RKHS space induced by the gradient of the infinite-width network at initialization. O is the standard O notation to get rid of constant.
3. Figure 2, see caption for the meaning of vertical coordinate. We verify Theorem 2.2 O(1/sqrt{n}) convergence rate.
4. It’s related to the learning parameter, the neural network architecture, and the maximum perturbation size we choose.
5. Our paper is not saying``SGD is robust to poisoning attacks’’. The whole point of the paper is to **identify conditions on the perturbations** under which SGD can efficiently and robustly learn important hypothesis classes such as linear models as well as two-layer neural networks.

Feel free to contact me via yunjuanwang97@gmail.com
